### ISSUE TITLE
Hadoop Shred: Lost events when task instances are used

### ISSUE THREAD

**Issue by bernardosrulzon** [snowplow/snowplow#3095]
Hey, guys!

We're doing quite a heavy debugging to understand why there's a significant (15-20%) difference between clicks reported by AdWords and page views reported by Snowplow. We figured out a good way of doing so:

- Redirect users to the target landing page using the collector redirect endpoint (`/r/tp2`). This helps us track clicks with the lowest possible friction - the number of `uri_redirects` should match the number of clicks very closely.
- To uniquely identify clicks, we can pass a `{random}` parameter into AdWords tracking template, which is set up as follows: `https://analytics.getninjas.com.br/r/tp2?u={escapedlpurl}?utm_source%3DGoogle%26utm_medium%3Dcpc%26utm_campaign%3D{campaignid}%26random%3D{random}`
- Since the `uri_redirect` event won't attach a `domain_userid`, the `{random}` parameter is what makes it possible to stitch a redirect with a page view, allowing us to analyze the funnel
- We will then compare the number of unique randoms on the `uri_redirect` table with the number of unique randoms on atomic.events - redirects should be greater than page views. Any differences here would be explained by (1) the client silently dropping the `utm` parameters, or (2) the load time being high enough that people give up before the page view gets sent to Snowplow.

Okay. Analyzing the data, this is what we found:

- We were surprised to find that 12% of randoms in atomic.events were actually not present in the redirect table, and investigated by looking at the usual suspects: raw, enrich and shred buckets
- The missing randoms were found on the raw logs (GET to /r/tp2, followed by the page_view POST) and the enriched/good files; but they did not show up in either shredded/good or shredded/bad.

We grepped the logs for one of the clicks (random = 15873185016763517746). **You can find them on [this gist](https://gist.github.com/bernardosrulzon/8fc25a5ac17c9418810de96ff6d4abda).**

**TL;DR**
12% of the redirects we expected to find on `uri_redirect` were correctly collected, enriched, but disappeared from the shred stage (neither good nor bad buckets).

Please let me know if you need more info to debug!

**Comment by alexanderdean** [comment]
Hi @bernardosrulzon - two things to check:

1. Is `continue_on_unexpected_error` set to true, and if so are there rows in the error bucket?
2. Are Hadoop Shred's deduplication algorithms responsible for the drop-off?

**Comment by bernardosrulzon** [comment]
Hey - thanks for the fast reply!

1. It's set to false
2. How can we check? It's probably not the case though, as we didn't see any events with that `random` in `uri_redirect`

**Comment by alexanderdean** [comment]
Right - if you are not seeing *any* events with that signature, then it's unlikely to be dedupe behind the drop off.

A good next step would be to plug one of the missing enriched events into the Hadoop Shred test suite and see if it makes it through or not. That will tell you if the problem is inside the Hadoop Shred process or outside it...

**Comment by bernardosrulzon** [comment]
@alexanderdean The event does get through the Hadoop Shred test suite. I'll try to re-process these events in our production pipeline to check whether this is an intermittent issue. Any other ideas?

Thanks!!

**Comment by alexanderdean** [comment]
Hi @bernardosrulzon - thanks for the update. I am not surprised by that discovery - but it is still puzzling that the events are going missing outside of the code path itself. Let us know what you find out!

**Comment by bernardosrulzon** [comment]
I think I found the culprit. I had recently updated the configs to use 1 core instance + 3 task instances, trying to lower EMR costs. It turns out the spot instances were responsible for the loss of events.

During this debugging, I used a fixed set of raw files and run the enrich/shred process with different configs. Using spot instances, I lost 5-15% of events - the lost events were random. I'm not sure if this problem affects other types of events - the case described above was the easiest for me to verify, so I only tested `uri_redirect`. Using core instances, all events came through.

For the record, the dataset had 546 unique events to be enriched and shredded. These were the experiments:
1. 1x `m3.2xlarge` for master, 1x `m3.2xlarge` for core, 3x `m3.2xlarge` for task (with bid = 0.532)
**Result:** 546 unique events in enrich (as expected). 523 unique events in shred (4.2% loss)

2. [Exactly the same as previous] 1x `m3.2xlarge` for master, 1x `m3.2xlarge` for core, 3x `m3.2xlarge` for task (with bid = 0.532)
**Result:** 546 unique events in enrich (as expected). 477 unique events in shred (12.6% loss). Lost events were different than previous run

2. [Removing task instances] 1x `m3.2xlarge` for master, 3x `m3.2xlarge` for core
**Result:** 546 unique events in enrich (as expected). 546 unique events in shred. Passed!

3. [Downsizing master] 1x `m1.medium` for master, 3x `m3.2xlarge` for core
**Result:** 546 unique events in enrich (as expected). 546 unique events in shred. Passed!

@alexanderdean Any ideas why this is happening and what the next steps would be?

**Comment by alexanderdean** [comment]
Hey @bernardosrulzon - this is really interesting, thanks for the detailed bug report.

Do you know if there was any volatility in your spot instances during the run - did you witness any of them being terminated/re-started?

Things we know:

* The Hadoop Shred job reads the enriched events for processing from HDFS
* The core instances do not get added to the HDFS cluster

Given this, it feels like the problem *might* be something like a spot instance been rebooted and then omitting some events when it resumes (i.e. "at-most-once processing"). Which would be a bug for sure.

But this is just speculation as to the cause - interested in whether others have experienced this?

**Comment by bernardosrulzon** [comment]
Good points. From the monitoring charts, the task instances took ~5min more to provision & bootstrap (vs. core). Otherwise, all other charts point to a healthy job - no reboots, no decommissions, no unhealthy nodes, etc. I set the bid pretty high exactly to avoid that.

**Comment by alexanderdean** [comment]
Interesting - even if the task instances take 5 minutes longer, they must be all settled by the time Hadoop Shred kicks off. Very puzzling...

**Comment by bernardosrulzon** [comment]
Exactly! Hadoop Shred kicks off only 22 mins into the job.

**Comment by alexanderdean** [comment]
What region is all this running in?

**Comment by bernardosrulzon** [comment]
N. Virginia (us-east-1d). The S3 buckets are in São Paulo, though.

**Comment by alexanderdean** [comment]
Need to try to reproduce this with Spark Shred...

**Comment by alexanderdean** [comment]
De-scheduling

**Comment by chuwy** [comment]
Hadoop Shred has been deprecated.
